sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+ */
+static const ble_uuid128_t gatt_svr_chr_stdin_uuid
+        = BLE_UUID128_INIT(0xaa, 0xf4, 0x82, 0xdd, 0x28, 0xa7, 0xac, 0x86, 0x68,
+                           0x4d, 0xd5, 0x40, 0x3f, 0x11, 0xdd, 0xcc);
+
+/**
+ * @brief Nimble access callback for stdin characteristic
+ */
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+void debug(const void* buffer, size_t len) {
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, len);
+}
+
+/**
+ * @brief Dummy access callback, because nimble requires one
+ */
+static int gatt_svr_chr_access_noop(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg) {
+
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) ctxt;
+    (void) arg;
+
+    return 0;
+}
+
+/**
+ * @brief Struct to define the stdio bluetooth service with its characteristics
+ */
+static const struct ble_gatt_svc_def _gatt_svr_svcs[] =
+{
+    /*
+     * access_cb defines a callback for read and write access events on
+     * given characteristics
+     */
+    {
+        /* Service: stdio */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = (ble_uuid_t*) &gatt_svr_svc_stdio_uuid.u,
+        .characteristics = (struct ble_gatt_chr_def[]) { {
+            /* Characteristic: stdout */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdout_uuid.u,
+            .access_cb = gatt_svr_chr_access_noop,
+            .val_handle = &_val_handle_stdout,
+            .flags = BLE_GATT_CHR_F_INDICATE,
+        }, {
+            /* Characteristic: stdin */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdin_uuid.u,
+            .access_cb = gatt_svr_chr_access_stdin,
+            .flags = BLE_GATT_CHR_F_WRITE,
+        }, {
+            0, /* No more characteristics in this service */
+        }, }
+    },
+    {
+        0, /* No more services */
+    },
+};
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg);
+
+static void _send_from_ringbuffer(bool mutex_locked);
+
+void stdio_init(void)
+{
+    ringbuffer_init(&_rb_stdin, _rb_stdin_mem, CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE);
+    ringbuffer_init(&_rb_stdout, _rb_stdout_mem, CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE);
+
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    ssize_t res;
+
+    /* block until at least on character was read from the buffer */
+    while (!(res = ringbuffer_get(&_rb_stdin, buffer, count))) {};
+
+    return res;
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    mutex_lock(&_rb_stdout_mutex);
+
+    unsigned int free = ringbuffer_get_free(&_rb_stdout);
+    unsigned int consumed = 0;
+
+    if (free > 0) {
+        unsigned int to_send = (free <= len) ? free : len;
+        consumed = ringbuffer_add(&_rb_stdout, buffer, to_send);
```

Comment:
I think we need to make it clear on what is the outcome on a write to a full buffer, here new data is lost, would it make more sense for old data do be lost instead? That clearing the buffer on a connect could be avoided (which means a memset), and we could only use static initializers

Reply:
I think it should work in FIFO fashion, so we only drop characters that do not fit anymore. I added a `@note` to the documentation of `stdio_nimble.h`. The buffer purge is now optional by a config value.

Reply:
I honestly don't have a super-strong opinion here, I'm just for harmonizing different stdio implementations. I'll ask some feedback around.